File: functions/lib/trends.js
"use strict";

function powerDictToSeries(dict) {
  const items = [];
  for (const [k, v] of Object.entries(dict || {})) {
    const val = Number(v);
    if (!Number.isFinite(val) || val <= -900) continue;
    const ks = String(k);
    if (ks.length < 8) continue;
    items.push({
      date: `${ks.slice(0, 4)}-${ks.slice(4, 6)}-${ks.slice(6, 8)}`,
      value: val,
    });
  }
  items.sort((a, b) => a.date.localeCompare(b.date));
  return items;
}

function linearTrend(values) {
  const n = values.length;
  if (n < 3) return { slope: 0, direction: "stable", r2: 0 };
  let sumX = 0, sumY = 0, sumXY = 0, sumXX = 0;
  for (let i = 0; i < n; i++) {
    sumX += i;
    sumY += values[i];
    sumXY += i * values[i];
    sumXX += i * i;
  }
  const denom = n * sumXX - sumX * sumX || 1;
  const slope = (n * sumXY - sumX * sumY) / denom;
  const intercept = (sumY - slope * sumX) / n;
  const mean = sumY / n;
  let ssRes = 0, ssTot = 0;
  for (let i = 0; i < n; i++) {
    const pred = intercept + slope * i;
    ssRes += (values[i] - pred) ** 2;
    ssTot += (values[i] - mean) ** 2;
  }
  const r2 = Math.max(0, Math.min(1, 1 - ssRes / (ssTot + 1e-9)));
  const rel = slope / (Math.abs(mean) + 1e-6);
  let direction = "stable";
  if (rel > 0.03) direction = "rising";
  else if (rel < -0.03) direction = "falling";
  return { slope, direction, r2 };
}

function forecastNext(values, horizon = 7) {
  if (!values.length) return Array(horizon).fill(0);
  if (values.length === 1) return Array(horizon).fill(values[0]);
  const n = values.length;
  let sumX = 0, sumY = 0, sumXY = 0, sumXX = 0;
  for (let i = 0; i < n; i++) {
    sumX += i; sumY += values[i]; sumXY += i * values[i]; sumXX += i * i;
  }
  const denom = n * sumXX - sumX * sumX || 1;
  const slope = (n * sumXY - sumX * sumY) / denom;
  const intercept = (sumY - slope * sumX) / n;
  let level = values[0];
  const alpha = 0.4;
  for (let i = 1; i < n; i++) level = alpha * values[i] + (1 - alpha) * level;
  const out = [];
  for (let i = 1; i <= horizon; i++) {
    const lin = intercept + slope * (n - 1 + i);
    out.push(0.55 * lin + 0.45 * level);
  }
  return out;
}

function explain(name, trend, unit) {
  const u = unit ? ` (${unit})` : "";
  if (trend.direction === "rising") return `${name} is trending up${u} — conditions are intensifying.`;
  if (trend.direction === "falling") return `${name} is trending down${u} — values are easing.`;
  return `${name} is relatively stable${u} over the recent window.`;
}

function analyzePowerBundle(power, horizon = 7) {
  const cfg = [
    ["rainfall", "Rainfall", "mm"],
    ["temp", "Temperature", "°C"],
    ["et", "Evapotranspiration", "mm"],
    ["humidity", "Humidity", "%"],
    ["solar", "Solar radiation", "kWh/m²"],
  ];
  const series = {};
  const trends = {};
  const forecasts = {};
  const explanations = [];

  for (const [key, label, unit] of cfg) {
    const pts = powerDictToSeries(power[key] || {});
    const vals = pts.map((p) => p.value);
    series[key] = pts;
    trends[key] = linearTrend(vals);
    forecasts[key] = forecastNext(vals, horizon);
    explanations.push(explain(label, trends[key], unit));
  }

  const rainDir = trends.rainfall?.direction;
  const etDir = trends.et?.direction;
  const tempDir = trends.temp?.direction;

  let outlook = {
    code: "stable",
    title: "Conditions look steady",
    message: "No strong shift detected over the next week based on recent trends.",
  };
  if (rainDir === "falling" && etDir === "rising") {
    outlook = {
      code: "drying",
      title: "Heading toward drier conditions",
      message:
        "Rainfall is easing while crop water demand (ET) is rising. Plan irrigation readiness over the next several days.",
    };
  } else if (rainDir === "rising") {
    outlook = {
      code: "wetting",
      title: "Heading toward wetter conditions",
      message: "Rainfall is increasing. Watch drainage and disease risk if humidity stays high.",
    };
  } else if (tempDir === "rising") {
    outlook = {
      code: "warming",
      title: "Heading toward warmer stress",
      message: "Temperatures are climbing. Heat stress risk may increase — consider cooling irrigation.",
    };
  }

  const predictedRain = (forecasts.rainfall || []).reduce((a, b) => a + b, 0);
  const predictedET = (forecasts.et || []).reduce((a, b) => a + b, 0);

  return {
    series,
    trends,
    forecasts,
    explanations,
    outlook: {
      ...outlook,
      predictedRain_mm: Math.round(predictedRain * 10) / 10,
      predictedET_mm: Math.round(predictedET * 10) / 10,
      horizonDays: horizon,
    },
  };
}

module.exports = { analyzePowerBundle, powerDictToSeries };

File: ml/trends.py
"""
Trend analysis + short-horizon forecast for field time series.

Works on NASA POWER daily series (rainfall, temp, ET, NDVI proxy, etc.).
"""

from __future__ import annotations

from typing import Dict, List, Optional

import numpy as np


def _series_from_power_dict(d: dict) -> List[dict]:
    """Convert POWER {YYYYMMDD: value} dict into sorted [{date, value}]."""
    items = []
    for k, v in (d or {}).items():
        try:
            val = float(v)
        except (TypeError, ValueError):
            continue
        if not np.isfinite(val) or val <= -900:  # POWER fill
            continue
        date = f"{str(k)[0:4]}-{str(k)[4:6]}-{str(k)[6:8]}"
        items.append({"date": date, "value": val})
    items.sort(key=lambda x: x["date"])
    return items


def linear_trend(values: List[float]) -> dict:
    y = np.asarray(values, dtype=np.float64)
    n = len(y)
    if n < 3:
        return {"slope": 0.0, "direction": "flat", "r2": 0.0}
    x = np.arange(n, dtype=np.float64)
    coef = np.polyfit(x, y, 1)
    slope = float(coef[0])
    pred = np.polyval(coef, x)
    ss_res = float(np.sum((y - pred) ** 2))
    ss_tot = float(np.sum((y - np.mean(y)) ** 2)) + 1e-9
    r2 = 1.0 - ss_res / ss_tot
    # normalize slope by mean magnitude
    scale = abs(float(np.mean(y))) + 1e-6
    rel = slope / scale
    if rel > 0.03:
        direction = "rising"
    elif rel < -0.03:
        direction = "falling"
    else:
        direction = "stable"
    return {"slope": slope, "direction": direction, "r2": float(max(0.0, min(1.0, r2)))}


def forecast_next(values: List[float], horizon: int = 7) -> List[float]:
    """Simple linear + exponential-smoothing hybrid forecast."""
    y = np.asarray(values, dtype=np.float64)
    if len(y) == 0:
        return [0.0] * horizon
    if len(y) == 1:
        return [float(y[0])] * horizon
    # linear
    x = np.arange(len(y), dtype=np.float64)
    slope, intercept = np.polyfit(x, y, 1)
    # exp smoothing level
    alpha = 0.4
    level = float(y[0])
    for v in y[1:]:
        level = alpha * float(v) + (1 - alpha) * level
    out = []
    for i in range(1, horizon + 1):
        lin = intercept + slope * (len(y) - 1 + i)
        # blend
        out.append(float(0.55 * lin + 0.45 * level))
    return out


def explain_trend(name: str, trend: dict, unit: str = "") -> str:
    d = trend["direction"]
    if d == "rising":
        return f"{name} is trending up{(' ('+unit+')') if unit else ''} — conditions are intensifying."
    if d == "falling":
        return f"{name} is trending down{(' ('+unit+')') if unit else ''} — values are easing."
    return f"{name} is relatively stable{(' ('+unit+')') if unit else ''} over the recent window."


def analyze_power_bundle(power: dict, horizon: int = 7) -> dict:
    """
    power: normalized POWER object with temp/rainfall/et/humidity/solar dicts.
    """
    metrics_cfg = [
        ("rainfall", "Rainfall", "mm"),
        ("temp", "Temperature", "°C"),
        ("et", "Evapotranspiration", "mm"),
        ("humidity", "Humidity", "%"),
        ("solar", "Solar radiation", "kWh/m²"),
    ]
    series = {}
    trends = {}
    forecasts = {}
    explanations = []

    for key, label, unit in metrics_cfg:
        pts = _series_from_power_dict(power.get(key) or {})
        vals = [p["value"] for p in pts]
        series[key] = pts
        tr = linear_trend(vals)
        trends[key] = tr
        forecasts[key] = forecast_next(vals, horizon=horizon)
        explanations.append(explain_trend(label, tr, unit))

    # Narrative "where it's leading"
    rain_dir = trends.get("rainfall", {}).get("direction")
    et_dir = trends.get("et", {}).get("direction")
    temp_dir = trends.get("temp", {}).get("direction")

    outlook = "stable"
    outlook_title = "Conditions look steady"
    outlook_message = "No strong shift detected over the next week based on recent trends."

    if rain_dir == "falling" and et_dir == "rising":
        outlook = "drying"
        outlook_title = "Heading toward drier conditions"
        outlook_message = (
            "Rainfall is easing while crop water demand (ET) is rising. "
            "Plan irrigation readiness over the next several days."
        )
    elif rain_dir == "rising" and (et_dir == "falling" or et_dir == "stable"):
        outlook = "wetting"
        outlook_title = "Heading toward wetter conditions"
        outlook_message = (
            "Rainfall is increasing. Watch drainage and disease risk if humidity stays high."
        )
    elif temp_dir == "rising":
        outlook = "warming"
        outlook_title = "Heading toward warmer stress"
        outlook_message = (
            "Temperatures are climbing. Heat stress risk may increase — consider cooling irrigation."
        )

    # Predicted 7-day totals
    pred_rain = float(np.sum(forecasts.get("rainfall") or [0]))
    pred_et = float(np.sum(forecasts.get("et") or [0]))

    return {
        "series": series,
        "trends": trends,
        "forecasts": forecasts,
        "explanations": explanations,
        "outlook": {
            "code": outlook,
            "title": outlook_title,
            "message": outlook_message,
            "predictedRain_mm": round(pred_rain, 1),
            "predictedET_mm": round(pred_et, 1),
            "horizonDays": horizon,
        },
    }
